crawler.go

```diff
@@ -329,21 +359,65 @@ func (c *Crawler) scanRequestWork(workCh chan chan *http.Request, closeCh chan i
 			spider.reqch <- requestAndChan{req: req, ch: resc}
 			select {
 			case re := <-resc:
-				closeRequest(req)
 				if re.err != nil {
 					c.logf("crawler: send HTTP request got error: %v", re.err)
 				} else {
-					go func(res *http.Response) {
+					go func(clonedReq *http.Request, res *http.Response) {
 						defer closeResponse(res)
 						defer func() {
 							if r := recover(); r != nil {
 								c.logf("crawler: Handler got panic error: %v", r)
 							}
 						}()
+
+						var retryResponseCode bool
+						if len(c.RetryHTTPResponseCodes) > 0 {
+							for _, v := range c.RetryHTTPResponseCodes {
+								if res.StatusCode == v {
+									retryResponseCode = true
+									break
+								}
+							}
+						}
+						url := clonedReq.URL.String()
+						var recrawl bool
+						c.urlNumRetriesMu.Lock()
+						if c.urlNumRetries == nil {
+							c.urlNumRetries = make(map[string]int)
+						}
+						if retryResponseCode {
+							val, _ := c.urlNumRetries[url]
+							if val < c.maxRetries() {
+								c.urlNumRetries[url]++
+								recrawl = true
+							}
+						} else {
+							// Finally, we succeeded! Reset the retry count.
+							c.urlNumRetries[url] = 0
```

Comment:
Should we delete it from the map instead of setting it to zero? Otherwise the map is ever growing, right?
i.e., `delete(c.urlNumRetries, url)`?

Reply:
good catch